op-service/sources/receipts.go

```diff
@@ -5,17 +5,277 @@ import (
+
+	// Fetch all receipts
+	for {
+		if err := call.Fetch(ctx); err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+	}
+	res, err := call.Result()
+	if err != nil {
+		return nil, err
+	}
+	// call successful, remove from cache
+	f.deleteBatchCall(block.Hash)
+	return res, nil
+}
+
+func (f *BasicRPCReceiptsFetcher) getOrCreateBatchCall(blockHash common.Hash, txHashes []common.Hash) *receiptsBatchCall {
+	f.callsMu.Lock()
+	defer f.callsMu.Unlock()
+	if call, ok := f.calls[blockHash]; ok {
+		return call
+	}
+	call := batching.NewIterativeBatchCall[common.Hash, *types.Receipt](
+		txHashes,
+		makeReceiptRequest,
+		f.client.BatchCallContext,
+		f.client.CallContext,
+		f.maxBatchSize,
+	)
+	f.calls[blockHash] = call
+	return call
+}
+
+func (f *BasicRPCReceiptsFetcher) deleteBatchCall(blockHash common.Hash) {
+	f.callsMu.Lock()
+	defer f.callsMu.Unlock()
+	delete(f.calls, blockHash)
+}
+
+type RPCReceiptsFetcher struct {
+	client rpcClient
+	basic  *BasicRPCReceiptsFetcher
+
+	log log.Logger
+
+	provKind RPCProviderKind
+
+	// availableReceiptMethods tracks which receipt methods can be used for fetching receipts
+	// This may be modified concurrently, but we don't lock since it's a single
+	// uint64 that's not critical (fine to miss or mix up a modification)
+	availableReceiptMethods ReceiptsFetchingMethod
+
+	// lastMethodsReset tracks when availableReceiptMethods was last reset.
+	// When receipt-fetching fails it falls back to available methods,
+	// but periodically it will try to reset to the preferred optimal methods.
+	lastMethodsReset time.Time
+
+	// methodResetDuration defines how long we take till we reset lastMethodsReset
+	methodResetDuration time.Duration
+}
+
+type RPCReceiptsConfig struct {
+	MaxBatchSize        int
+	ProviderKind        RPCProviderKind
+	MethodResetDuration time.Duration
+}
+
+func NewRPCReceiptsFetcher(client rpcClient, log log.Logger, config RPCReceiptsConfig) *RPCReceiptsFetcher {
+	return &RPCReceiptsFetcher{
+		client:                  client,
+		basic:                   NewBasicRPCReceiptsFetcher(client, config.MaxBatchSize),
+		log:                     log,
+		provKind:                config.ProviderKind,
+		availableReceiptMethods: AvailableReceiptsFetchingMethods(config.ProviderKind),
+		lastMethodsReset:        time.Now(),
+		methodResetDuration:     config.MethodResetDuration,
+	}
+}
+
+func NewCachingRPCReceiptsProvider(client rpcClient, log log.Logger, config RPCReceiptsConfig, m caching.Metrics, cacheSize int) *CachingReceiptsProvider {
+	return NewCachingReceiptsProvider(NewRPCReceiptsFetcher(client, log, config), m, cacheSize)
+}
+
+func (f *RPCReceiptsFetcher) FetchReceipts(ctx context.Context, block eth.BlockID, txHashes []common.Hash) (result types.Receipts, err error) {
+	m := f.PickReceiptsMethod(len(txHashes))
+	switch m {
+	case EthGetTransactionReceiptBatch:
+		result, err = f.basic.FetchReceipts(ctx, block, txHashes)
```

Comment:
Out of scope for this PR (and maybe not worth it), but I noticed that this basic batch get is the only one we have a unique method/struct for & the rest we are doing raw CallContexts for.